### ISSUE TITLE
GRPO: For the same batch of data, there is a significant difference in model performance between using trainer. predict() and using trainer. train().

### ISSUE THREAD

**Issue by yzhdut** [huggingface/trl#2970]
This problem has been bothering me for a long time, and I would be grateful if someone could help me. I found that when I started training, the model performed poorly and the reward remained the lowest, so I doubted if it could be trained successfully. But when I used trainer. predict() to test the same batch of data, I found that the effect was much better than when training. I'm confused, is this normal?

env：
CUDA Version: 12.4
torch 2.0.0+cu118
trl 0.15.1
transformers 4.46.0

reference code：
`
with open('data/RLFTqa_delete_some_answer_is_null.json', 'r') as f:
data = json.load(f)
random.shuffle(data)
data = data[0:200]
formatted_data = []
for item in data:
formatted_data.append({
"prompt": item["question"],
"reference": "||".join(item.get("answer", []))
})

n_total = len(formatted_data)
n_train = int(0.7 * n_total)
n_eval = int(0.15 * n_total)
n_test = n_total - n_train - n_eval

train_data = formatted_data[:1]
eval_data = formatted_data[n_train:n_train+n_eval]
test_data = formatted_data[n_train+n_eval:]

dataset_features = Features({
"prompt": Value("string"),
"reference": Value("string")
})

train_dataset = Dataset.from_list(train_data, features=dataset_features)
eval_dataset = Dataset.from_list(eval_data, features=dataset_features)
test_dataset = Dataset.from_list(test_data, features=dataset_features)

local_model_path = "./Qwen2.5-7B-Instruct-GPTQ-Int4"
tokenizer = AutoTokenizer.from_pretrained(local_model_path)
tokenizer.pad_token = tokenizer.eos_token

base_model = AutoModelForCausalLM.from_pretrained(
local_model_path,
# quantization_config=quant_config,
device_map="auto",
trust_remote_code=True,
torch_dtype=torch.float16
)
base_model.config.gradient_checkpointing = False

lora_config = LoraConfig(
r=16, # 适当增加秩维度
lora_alpha=32,
target_modules=["q_proj", "k_proj", "v_proj", "o_proj"], # 扩展目标模块
lora_dropout=0.1,
bias="none",
task_type="CAUSAL_LM",
)

model = prepare_model_for_kbit_training(base_model)
model = get_peft_model(model, lora_config)

grpo_config = GRPOConfig(
num_generations=8, # 每组生成数量
max_completion_length=512, # 最大生成长度
gradient_accumulation_steps=2,
learning_rate=3e-5,

logging_steps=5,
logging_first_step = True,
save_steps=50,
output_dir="./grpo_checkpoints",
logging_dir="./grpo_checkpoints/log",
log_level="info",
log_completions=True,

eval_strategy="steps",
eval_steps=40,

do_predict=True,

temperature=0.7,
gradient_checkpointing=False,
use_vllm=False
)
model.bfloat16()
trainer = GRPOTrainer(
model=model,
args=grpo_config,
reward_funcs=[reward_func, format_reward_func], # 直接传递奖励函数
train_dataset=train_dataset,
eval_dataset=eval_dataset,
processing_class=tokenizer,
)
train_output = trainer.predict(train_dataset)

trainer.train()

model.save_pretrained("grpo_lora_adapter")`

As mentioned above, the length of train_det is 1. Before training, I first call trainer. predict (train_dataset). Then I printed the completion of the model in the library function, and the obtained location is as follows:
xxx/lib/python3.10/site-packages/trl/trainer/grpo_trainer.py：
` def _prepare_inputs(self, inputs: dict[str, Union[torch.Tensor, Any]]) -> dict[str, Union[torch.Tensor, Any]]:
device = self.accelerator.device
prompts = [x["prompt"] for x in inputs]
prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
prompt_inputs = self.processing_class(
prompts_text, return_tensors="pt", padding=True, padding_side="left", add_special_tokens=False
)
prompt_inputs = super()._prepare_inputs(prompt_inputs)
prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
if self.max_prompt_length is not None:
prompt_ids = prompt_ids[:, -self.max_prompt_length :]
prompt_mask = prompt_mask[:, -self.max_prompt_length :]

    # Generate completions using either vLLM or regular generation
    if self.args.use_vllm:
        print("Using VLM model")
        # First, have main process load weights if needed
        if self.state.global_step != self._last_loaded_step:
            self._move_model_to_vllm()
            self._last_loaded_step = self.state.global_step

        # Generate completions using vLLM: gather all prompts and use them in a single call in the main process
        all_prompts_text = gather_object(prompts_text)
        if self.accelerator.is_main_process:
            outputs = self.llm.generate(all_prompts_text, sampling_params=self.sampling_params, use_tqdm=False)
            completion_ids = [out.token_ids for completions in outputs for out in completions.outputs]
        else:
            completion_ids = [None] * len(all_prompts_text)
        # Broadcast the completions from the main process to all processes, ensuring each process receives its
        # corresponding slice.
        completion_ids = broadcast_object_list(completion_ids, from_process=0)
        process_slice = slice(
            self.accelerator.process_index * len(prompts),
            (self.accelerator.process_index + 1) * len(prompts),
        )
        completion_ids = completion_ids[process_slice]

        # Pad the completions, and concatenate them with the prompts
        completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
        completion_ids = pad(completion_ids, padding_value=self.processing_class.pad_token_id)
        prompt_completion_ids = torch.cat([prompt_ids, completion_ids], dim=1)
    else:
        # Regular generation path
        with unwrap_model_for_generation(self.model, self.accelerator) as unwrapped_model:
            prompt_completion_ids = unwrapped_model.generate(
                prompt_ids, attention_mask=prompt_mask, generation_config=self.generation_config
            )

        # Compute prompt length and extract completion ids
        prompt_length = prompt_ids.size(1)
        prompt_ids = prompt_completion_ids[:, :prompt_length]
        completion_ids = prompt_completion_ids[:, prompt_length:]

    # Mask everything after the first EOS token
    is_eos = completion_ids == self.processing_class.eos_token_id
    eos_idx = torch.full((is_eos.size(0),), is_eos.size(1), dtype=torch.long, device=device)
    eos_idx[is_eos.any(dim=1)] = is_eos.int().argmax(dim=1)[is_eos.any(dim=1)]
    sequence_indices = torch.arange(is_eos.size(1), device=device).expand(is_eos.size(0), -1)
    completion_mask = (sequence_indices <= eos_idx.unsqueeze(1)).int()

    # Concatenate prompt_mask with completion_mask for logit computation
    attention_mask = torch.cat([prompt_mask, completion_mask], dim=1)  # (B*G, P+C)

    logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens

    with torch.inference_mode():
        if self.ref_model is not None:
            ref_per_token_logps = self._get_per_token_logps(
                self.ref_model, prompt_completion_ids, attention_mask, logits_to_keep
            )
        else:
            with self.accelerator.unwrap_model(self.model).disable_adapter():
                ref_per_token_logps = self._get_per_token_logps(
                    self.model, prompt_completion_ids, attention_mask, logits_to_keep
                )

    # Decode the generated completions
    completions_text = self.processing_class.batch_decode(completion_ids, skip_special_tokens=True)
    print("Completions text:", completions_text)`
As I mentioned, when predicting, the model shows basic problem-solving ability. However, for the same input, during training, completion appears to significantly reduce model performance due to issues with irrelevant statements and tasks, as well as repetition.

And I found that the log doesn't seem to be saved together with the checkpoint. Is it possible that it will only be saved after the training is completed

**Comment by lyh1028** [comment]
have the same question

**Comment by emaadmanzoor** [comment]
Following

**Comment by albertvillanova** [comment]
Hey @yzhdut, sorry for the very late reply — we're catching up on the backlog. 

This is expected behavior rather than a bug: GRPO generates with sampling (`do_sample=True`, `temperature=0.7`, see https://github.com/huggingface/trl/blob/fe19a853d7201e00f4c6ff86e09ae4613bb52821/trl/trainer/grpo_trainer.py#L893-L900) on *both* the `train()` and `predict()` paths, so completions for the same prompt naturally differ run-to-run — nicer outputs in one `predict()` call than in a training step is just sampling variance. Two things also work against you here: you're training on a single example (`train_data = formatted_data[:1]`), and with one prompt per group GRPO normalizes rewards within the group, so the advantages collapse to ~0 and there's essentially no learning signal — that's why the reward stays flat. The setup also mixes a couple of anti-patterns (`prepare_model_for_kbit_training` on a model loaded without a quantization config, then `model.bfloat16()` after a `float16` load) that can degrade generations on their own. If you want train/eval to line up more closely, current TRL exposes `disable_dropout=True` in `GRPOConfig`. 

I'll close this as not-a-bug, but I'm happy to re-open if you can share a minimal, runnable repro (small public dataset + model) that shows a genuine TRL discrepancy.
